core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -376,12 +381,36 @@ class Partition(val topic: String,
   }
 
   /**
+   * Check and maybe increment the low watermark of the partition;
+   * this function can be triggered when any replica's logStartOffset changed
```

Comment:
Yes we can compute it on the fly when we need it in e.g. in `DelayedDeleteRecords`. Is there reason to do it on the fly instead of maintaining it in the `Replica`? Is it because you think the patch would be simpler if we compute it on the fly?

I think both solution has the same performance. But maintaining it in Replica in the same way as `HW` may make LW easier to reason about and extend in the future. For example, the code is easier to extend if we want to use `LW` in the future as a way to determine the minimum offset available for consumption (currently the minimum offset available for consumption varies between replicas because different replicas may have different logStartOffset).